File: TinyScript/unstable/python/lexer/Lexer.py
# -*- coding: UTF-8 -*-
'''=================================================
@Project -> File   ：TinyScript -> Lexer
@IDE    ：PyCharm
@Author ：QiangZiBro
@Date   ：2020/4/12 6:10 下午
@Desc   ：
=================================================='''

from common.PeekIterator import PeekIterator
from .AlphabetHelper import AlphabetHelper
from .Token import makeNumber, makeOperator, makeString, makeVarOrKeyword, Token
from lexer import TokenType
from lexer.LexicalException import LexicalException


class Lexer:
    def analyse(self, source):
        """

        :param source: code iterator
        :return:
        """
        tokens = []
        it = PeekIterator(source, "\0")
        helper = AlphabetHelper()
        while it.hasNext():
            c = it.next()
            if c == "\0":
                break
            lookahead = it.peek()
            # delete comments
            if c == "/":
                if lookahead == "/":
                    c = it.next()
                    while it.hasNext() and c != "\n":
                        c = it.next()
                elif lookahead == "*":
                    valid = False
                    while it.hasNext():
                        p = it.next()
                        if p == "*" and it.peek() == "/":
                            valid = True
                            it.next()
                            break
                    if not valid:
                        raise LexicalException("comment not matched")
                    continue

            # brackets
            if c in "{}()":
                tokens.append(Token(TokenType.BRACKET, c))
                continue

            # number
            # case1: 334+...
            if helper.isNumber(c):
                it.putBack()
                tokens.append(makeNumber(it))
                continue
            # case2: +5, 3*-5, where c is +,- respectively
            if c in "+-" and helper.isNumber(lookahead):
                # jump 3 + 5, a + 5
                if len(tokens) == 0:
                    lastToken = None
                else:
                    lastToken = tokens[-1]

                # two cases that can think it a number +5, 3*-5
                if lastToken is None or not lastToken.isValue():
                    it.putBack()
                    token = makeNumber(it)
                    tokens.append(token)
                    continue
            # operator
            if helper.isOperator(c):
                it.putBack()
                token = makeOperator(it)
                tokens.append(token)
                continue

            # string
            if c == "'" or c == '"':
                it.putBack()
                token = makeString(it)
                tokens.append(token)
                continue

            # variable or keyword
            if helper.isLiteral(c):
                it.putBack()
                token = makeVarOrKeyword(it)
                tokens.append(token)
        return tokens

File: TinyScript/unstable/typescript/src/lexer/Lexer.ts
import { Token, TokenType } from "./Token";
import PeekIterator from "../common/PeekIterator";
import arrayToGenerator from "../common/arrayToGenerator";
import LexicalException from "./LexicalException";
import AlphabetHelper from "./AlphabetHelper";
import fs from 'fs'

/**
 * 词法分析
 * @input source
 * @output Array<Token>
 */
class Lexer {
  // 关键点：生成Token的有限状态机
  analyse(source: string): Array<Token<string>> {
    const tokens: Array<Token<string>> = []
    const it = new PeekIterator(arrayToGenerator([...source]), '\0')

    while (it.hasNext()) {
      let c = it.next()
      if (c === '\0') {
        break
      }
      let lookahead = it.peek()
      // 跳过空字符
      if (c === ' ' || c === '\n' || c === '\r') {
        continue
      }
      // 提取注释的程序
      if (c === '/') {
        if (lookahead === '/') {
          while (it.hasNext() && (c = it.next()) !== '\n');
          continue
        } else if (lookahead === '*') {
          let valid = false;
          while (it.hasNext()) {
            const p = it.next();
            if (p === "*" && it.peek() === "/") {
              valid = true;
              it.next();
              break;
            }
          }
          if (!valid) {
            throw new LexicalException("comment not matched");
          }
          continue;
        }
      }
      if (c == "{" || c == "}" || c == "(" || c == ")") {
        tokens.push(new Token(TokenType.BRACKET, c))
        continue
      }
      if (c == '"' || c == "'") {
        it.putBack();
        tokens.push(Token.makeString(it));
        continue;
      }

      if (AlphabetHelper.isLetter(c)) {
        it.putBack();
        tokens.push(Token.makeVarOrKeyword(it));
        continue;
      }

      if (AlphabetHelper.isNumber(c)) {
        it.putBack();
        tokens.push(Token.makeNumber(it));
        continue;
      }

      // + -
      if ((c == "+" || c == "-") && AlphabetHelper.isNumber(lookahead)) {
        // 跳过:a+1, 1+1
        // +5, 3*-5
        const lastToken = tokens[tokens.length - 1] || null;
        if (lastToken == null || !lastToken.isValue()) {
          it.putBack();
          tokens.push(Token.makeNumber(it));
          continue;
        }
      }

      if (AlphabetHelper.isOperator(c)) {
        it.putBack();
        tokens.push(Token.makeOp(it));
        continue;
      }

      throw LexicalException.fromChar(c);
    }
    return tokens
  }

  static fromFile(src: string): Array<Token<string>> {
    const content = fs.readFileSync(src, "utf-8")
    const lexer = new Lexer()
    return lexer.analyse(content)
  }
}

export default Lexer